### ISSUE TITLE
Replication issue

### ISSUE THREAD

**Issue by ChoiHongrok** [levtelyatnikov/EGG_GAE#1]
Hi, Thanks for sharing your code.
I am currently researching a subject similar to yours and want to replicate your EGG-GAE model.
However, I obtained different model results using your GitHub code compared to the results presented in your paper.
Could you please clarify any aspects I might have misunderstood or any details not thoroughly described?
The experimental settings and methods are below.
I appreciate your assistance and look forward to your reply. Thank you

I used 3 models, EGG-GAE, GAIN, and MICE

Dataset: "Adult" in your Github

[model 1] K-EGG-GAE
1) Modify defaults.yaml and dataloader.yaml to fit the Adult dataset.
2) Train the model using the [run.py](http://run.py/) file and save the checkpoints (5 runs)
3) make predictions on the test set and measure RMSE and Accuracy (the attached file)
4) Results(MCAR//MAR/MNAR) - RMSE:
   - Paper results: : 0.8538±0.0032 // 0.8323±0.0017 // 0.8639±0.0043
   - My experimental results: 0.443±0.003 // 0.399±0.003 // 0.431±0.001
5) Results(MCAR//MAR/MNAR) - Accuracy:
   - Paper results: 72.51±0.34 // 73.35±0.55 // 73.68±0.41
   - My experimental results: 0.912±0.002 // 0.902±0.001 // 0.911±0.002
  
![image](https://github.com/levtelyatnikov/EGG_GAE/assets/57215124/e9051c22-4e2f-499e-bc8b-365c176ab325)
![image](https://github.com/levtelyatnikov/EGG_GAE/assets/57215124/8881ab25-3611-4e7f-b11a-db414b75bdcd)
```python
from glob import glob
import torch
dev = torch.device('cuda')
from hydra import initialize, compose
from utils import config_preprocess, get_dataloader
from method import LitModel

import numpy as np
import pandas as pd
from tqdm import tqdm
            
class dotdict(dict):
    """dot.notation access to dictionary attributes"""
    __getattr__ = dict.get
    __setattr__ = dict.__setitem__
    __delattr__ = dict.__delitem__
    
# for dataset in ['credit', 'adult']:
for dataset in ['Adult']:
    for mech in ['mcar', 'mar', 'mnar']:
        PATHs = sorted(glob(f'outputs/{dataset}_{mech}/*/*/'))
        for PATH in PATHs:
            print(PATH)
            idx = PATH.split('/')[-2]
            with initialize(version_base=None, config_path=PATH+'.hydra'):
                cfg = compose(config_name=f'config.yaml')

            datamodule = get_dataloader(cfg)
            cfg = config_preprocess(cfg, datamodule)
            data_name = cfg.dataloader.dataset_name
            miss_algo = cfg.dataloader.miss_algo
            print(data_name, miss_algo)
            model = LitModel(datamodule=datamodule, cfg=cfg)

            # 모델에 불러오기
            PATH_model = sorted(glob(PATH+'checkpoints/*'))[-1]
            print(PATH_model)
            loaded = torch.load(PATH_model)
            state_dict = {key[6:] : val for key, val in loaded['state_dict'].items()}
            model.model.to(dev)
            model.model.load_state_dict(state_dict)

            X_clean = np.empty(datamodule.dl.test_dataset.features_freq.shape)
            X_imp = np.empty(datamodule.dl.test_dataset.features_freq.shape)
            X_mask = np.empty(datamodule.dl.test_dataset.features_freq.shape)
            for dataloader, mode in zip([datamodule.test_dataloader()], 
                                        ['test', 'train']):
                probs, preds, y_test, X_test, X_test_rec, masks = [], [], [], [], [], []
                unique_set = set()
                for batch in tqdm(dataloader, total=len(dataloader)):
                    prob, pred, y, rec_x = model.model.inference(batch.x.to(dev), batch.y.to(dev), batch)
                    unique_set.update(batch.unique_idx.cpu().numpy())
                    X_imp[batch.unique_idx] = rec_x.cpu().detach().numpy() 
                    X_clean[batch.unique_idx] = batch.x_clean.cpu().detach().numpy() 
                    X_mask[batch.unique_idx] = batch.mask.cpu().detach().numpy()
                    if len(unique_set) == len(X_imp):
                        break
                X_imp[X_mask.astype(bool)] = X_clean[X_mask.astype(bool)]
            np.save(f'/home/redgreen/ssl/metric/egg_gae/{dataset}_{mech}_{idx}', X_imp)
```

[Model 2] GAIN
1) Use the dataset made by K-EGG-GAE training process
2) Use both the train and test sets for prediction
3) Code: https://github.com/jsyoon0823/GAIN
4) Results(MCAR//MAR/MNAR) - RMSE:
   - Paper results: 0.9983±0.0063 // 1.514±0.5263 // 1.0271±0.0023
   - My experimental results: 1.734±0.694 // 0.583±0.13 // 0.756±0.063
5) Results(MCAR//MAR/MNAR) - Accuracy:
   - Paper results:  29.9±4.27 // 22.46±4.15 // 35.39±3.78
   - My experimental results:  0.86±0.011 // 0.865±0.007 // 0.853±0.01

```python
import os 

os.environ["CUDA_DEVICE_ORDER"]="PCI_BUS_ID"  # Arrange GPU devices starting from 0
os.environ["CUDA_VISIBLE_DEVICES"]= "7"  # Set the GPUs 2 and 3 to use

import numpy as np
from gain import gain
from utils import rmse_loss
from data_loader import data_loader

def make_dataset(data, mode):
    X = data[f'X_{mode}_clean']
    mask = data[f'mask_init_{mode}']
    X_mis = X.copy()
    X_mis[~mask.astype(bool)] = np.nan
    return X, X_mis, mask

def merge_X(X_train, X_test_mis):
    X_merge = np.concatenate([X_train, X_test_mis], axis=0)
    return X_merge 
    
gain_parameters = {'batch_size': 128,
                    'hint_rate': 0.9,
                    'alpha': 100,
                    'iterations': 10000,
                    'cat_ids' : None}
# preds = {}

for dataset in ['Adult']:
# for dataset in ['adult', 'credit']:
    for miss_algo in ['mcar', 'mar', 'mnar']:
        npz = np.load(f'/home/redgreen/ssl/EGG_GAE/noised_datasets/{dataset}_{miss_algo}/{dataset}/{dataset}.npz', allow_pickle=True)
        # npz = np.load(f'/home/redgreen/ssl/EGG_GAE/noised_datasets/{dataset}_{miss_algo}_us/{dataset}_us/{dataset}_us.npz', allow_pickle=True)
        gain_parameters['cat_ids'] = npz['cat_idx']
        X_train, X_train_mis, mask_train = make_dataset(npz, mode='train')
        X_test, X_test_mis, mask_test = make_dataset(npz, mode='test')
        X_merge = merge_X(X_train, X_test_mis)
        for i in range(5):
            imputed_data_x = gain(X_merge, gain_parameters)
            # preds[miss_algo] = imputed_data_x
            np.save(f'/home/redgreen/ssl/metric/gain/{dataset}_{miss_algo.lower()}_{i}', imputed_data_x[len(X_train):])
```
[Model 3] MICE
1) Use the dataset made by K-EGG-GAE training process
2) Use both the train and test sets for prediction
3) Code: MICE library in R
4) Results(MCAR//MAR/MNAR) - RMSE:
   - Paper results: 0.9574±0.0//0.9779±0.0//0.9772±0.0
   - My experimental results: 1.097±0 // 0.389±0 // 0.372±0
5) Results(MCAR//MAR/MNAR) - Accuracy:
   - Paper results: 19.64±0.0 // 20.67±0.0 // 22.9±0.0
   - My experimental results: 0.921±0 // 0.914±0 // 0.921±0
```R
library(reticulate)
library(hash)
library(mice)
#py_install('numpy')
np <- import('numpy')

mechs <- c('mcar', 'mar', 'mnar')
datasets <- c('Adult', 'credit', 'adult')

# feature name
col_names <- hash()
col_names[['adult']] <- c('age', 'workclass', 'fnlwgt', 'education', 'education_num',
                          'marital_status', 'occupation', 'relationship', 'race', 'sex',
                          'capital_gain', 'capital_loss', 'hours_per_week', 'native_country')
col_names[['Adult']] <- c('age', 'workclass', 'fnlwgt', 'education', 'education_num',
                          'marital_status', 'occupation', 'relationship', 'race', 'sex',
                          'capital_gain', 'capital_loss', 'hours_per_week', 'native_country')
col_names[['credit']] <- c('LIMIT_BAL', 'SEX', 'EDUCATION', 'MARRIAGE', 'AGE', 
                           'PAY_0', 'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 
                           'BILL_AMT1', 'BILL_AMT2', 'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 
                          'PAY_AMT1', 'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6')
# feature type
cat_ids <- hash()
#cat_ids[['adult']] <- c(1, 3, 5, 6, 7, 8, 9, 13) + 1
cat_ids[['adult']] <- c(1, 3, 5, 6, 7, 8, 9) + 1 
cat_ids[['Adult']] <- c(1, 3, 5, 6, 7, 8, 9) + 1 
#native_country를 factor로 지정하면 class의 개수가 너무 많아서(43개) 오류가남 어쩔 수 없이 pmm으로 사용
cat_ids[['credit']] <- c(1, 2, 3) + 1

###############################################
################## Functions ##################
###############################################
extract.data <- function(npz){
  X_train <- npz[['X_train_clean']]
  n_train <- dim(X_train)[1]
  X_test <- npz[['X_test_clean']]
  X <- rbind(X_train, X_test)
  mask_train <- npz[['mask_init_train']]
  mask_train <- matrix(1, ncol=ncol(mask_train), nrow=nrow(mask_train))
  mask_test <- npz[['mask_init_test']]
  mask <- rbind(mask_train, mask_test)
  X <- data.frame(X)
  mask <- data.frame(mask)
  
  # change feature type
  for (c_i in cat_ids[[dataset]]){
    X[, c_i] = as.factor(X[, c_i])
  }
  # set feature name
  colnames(X) <- col_names[[dataset]]
  X_mis <- replace(X, mask==0, NA)
  
  list(X=X, mask=mask, X_mis=X_mis, n_train=n_train)
}

getmode <- function(v){
  uniqv <- unique(v)
  uniqv[which.max(tabulate(match(v, uniqv)))]
}

pool.impute <- function(X_mis, mask, imp, cat_ids){
  for (i in 1:length(X_mis)){
    if (i %in% cat_ids){
      fun = getmode
    }else{
      fun = median
    }
    colname = colnames(X_mis)[i]
    sampled <- apply(imp$imp[[colname]], 1, fun)
    X_mis[, i] <-replace(X_mis[,i], mask[,i]==0, sampled)
  }
  X_mis
}
###############################################
###############################################
###############################################

###############################################
#################### MAIN #####################
###############################################
for (dataset in datasets){
  for (mech in mechs){
    path <- paste0('noised_datasets/',dataset,'_',mech,'/',dataset,'/',dataset,'.npz')
    if (dataset == "Adult"){
      path <- paste0('noised_datasets/_',dataset,'_',mech,'/',dataset,'/',dataset,'.npz')
    }
    npz <- np$load(path, allow_pickle = TRUE)
    data <- extract.data(npz)
    X <- data[['X']]
    mask <- data[['mask']]
    X_mis <- data[['X_mis']]
    n_train <- data[['n_train']]
    
    imp <- mice(X_mis, m=20)
    X_imp <- pool.impute(X_mis, mask, imp, cat_ids[[dataset]])
    if (dataset == "Adult"){
      np$save(paste0('mice/',mech,'__',dataset,'_mice.npy'), as.matrix(X_imp)[(n_train+1):nrow(X_imp),])
    }else{
      np$save(paste0('mice/',mech,'_',dataset,'_mice.npy'), as.matrix(X_imp)[(n_train+1):nrow(X_imp),])
    }
  
  }
}
###############################################
###############################################
```

**Comment by levtelyatnikov** [comment]
Hello, thank you for sharing this issue!

As far as I can see, the results for k-EGG-GAE improved compared to the paper's results. Have you modified the data preprocessing in any way? If you want, we can set up a meeting to figure out the discrepancy.

**Comment by ChoiHongrok** [comment]
No, i didn't modify the data preprocessing. I simply used the code you provided. 
To use trained model setting, I loaded `checkpoint(ckpt) file` and `config.yaml file` from `outputs folder`.
And I used `datamodule.test_dataloader()` method for making prediction by batch.
I checked `datamodule.test_dataloader()` has duplicated idx, so i use unique idx rows for evaluation.
I think there isn't anything strange. 
Please let me know if there is anything you think is strange or suspicious.
